sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Average.scala

```diff
@@ -46,6 +46,8 @@ case class Average(
 
   def this(child: Expression) = this(child, failOnError = SQLConf.get.ansiEnabled)
 
+  override lazy val resolved: Boolean = children.forall(_.resolved)
```

Comment:
Before this change, this `resolved` field of Average is set based on both children all resolved and input data type match (the default Expression implementation). However this will lead to un-resolved column name with Average in HAVING due to the logic of TempResolveColumn, which lead to column not found (but the column was found).

If we set this field only based on if all children are resolved, later in `CheckAnalysis` it will check expression input data type, and throw right `data type mismatch` error: https://github.com/apache/spark/blob/0d56c947f10f747ab4b76426b2d6a34a1d3b8277/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L189

Reply:
It's intentional that `Expression.resolved` also checks the input data types. It seems weird to me that we only change it for `Average`...

Reply:
Do you agree the way in this PR for Average? If so I can go to check existing expressions and make them follow the same way?

Basically if we make one expression as resolved when `all children is solved` && `data type match`, in the case of the TempResolveColumn usage, we could have wrong error message in the later stage (it reports `column not found`, which is not true).

If we change to say such expression can be marked as resolved only `all children is solved`, later in the CheckAnalysis, it will report data mismatch. 

If the proposal change is not the good way to fix this issue, is there a better way?